In [4]:
import graphlab as gl

In [48]:
train_data = gl.SFrame("../data/kaggle_titanic/train.csv")
test_data = gl.SFrame("../data/kaggle_titanic/test.csv")

Finished parsing file /Users/danolee/python/ml_training/hyundai card/data/kaggle_titanic/train.csv

Parsing completed. Parsed 100 lines in 0.047524 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[int,int,int,str,str,float,int,int,str,float,str,str]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Finished parsing file /Users/danolee/python/ml_training/hyundai card/data/kaggle_titanic/train.csv

Parsing completed. Parsed 891 lines in 0.012802 secs.

Finished parsing file /Users/danolee/python/ml_training/hyundai card/data/kaggle_titanic/test.csv

Parsing completed. Parsed 100 lines in 0.014995 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[int,int,str,str,float,int,int,str,float,str,str]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Finished parsing file /Users/danolee/python/ml_training/hyundai card/data/kaggle_titanic/test.csv

Parsing completed. Parsed 418 lines in 0.013075 secs.

# SFrame에서 특정 컬럼만 선택하기

    SFrame[컬럼이름들]

In [52]:
train_data.head()

Columns:
	PassengerId	int
	Survived	int
	Pclass	int
	Name	str
	Sex	str
	Age	float
	SibSp	int
	Parch	int
	Ticket	str
	Fare	float
	Cabin	str
	Embarked	str
	Embarked_filled	str

Rows: 10

Data:
+-------------+----------+--------+--------------------------------+--------+
| PassengerId | Survived | Pclass |              Name              |  Sex   |
+-------------+----------+--------+--------------------------------+--------+
|      1      |    0     |   3    |    Braund, Mr. Owen Harris     |  male  |
|      2      |    1     |   1    | Cumings, Mrs. John Bradley...  | female |
|      3      |    1     |   3    |     Heikkinen, Miss. Laina     | female |
|      4      |    1     |   1    | Futrelle, Mrs. Jacques Hea...  | female |
|      5      |    0     |   3    |    Allen, Mr. William Henry    |  male  |
|      6      |    0     |   3    |        Moran, Mr. James        |  male  |
|      7      |    0     |   1    |    McCarthy, Mr. Timothy J     |  male  |
|      8      |    0     |   3    | Palsson, Master. Gosta Leonard |  male  |
|      9      |    1     |   3    | Johnson, Mrs. Oscar W (Eli...  | female |
|      10     |    1     |   2    | Nasser, Mrs. Nicholas (Ade...  | female |
+-------------+----------+--------+--------------------------------+--------+
+------+-------+-------+------------------+---------+-------+----------+-----------------+
| Age  | SibSp | Parch |      Ticket      |   Fare  | Cabin | Embarked | Embarked_filled |
+------+-------+-------+------------------+---------+-------+----------+-----------------+
| 22.0 |   1   |   0   |    A/5 21171     |   7.25  |       |    S     |        S        |
| 38.0 |   1   |   0   |     PC 17599     | 71.2833 |  C85  |    C     |        C        |
| 26.0 |   0   |   0   | STON/O2. 3101282 |  7.925  |       |    S     |        S        |
| 35.0 |   1   |   0   |      113803      |   53.1  |  C123 |    S     |        S        |
| 35.0 |   0   |   0   |      373450      |   8.05  |       |    S     |        S        |
| None |   0   |   0   |      330877      |  8.4583 |       |    Q     |        Q        |
| 54.0 |   0   |   0   |      17463       | 51.8625 |  E46  |    S     |        S        |
| 2.0  |   3   |   1   |      349909      |  21.075 |       |    S     |        S        |
| 27.0 |   0   |   2   |      347742      | 11.1333 |       |    S     |        S        |
| 14.0 |   1   |   0   |      237736      | 30.0708 |       |    C     |        C        |
+------+-------+-------+------------------+---------+-------+----------+-----------------+
[10 rows x 13 columns]

### 컬럼 하나 선택하기

In [53]:
train_data['Fare']

dtype: float
Rows: 891
[7.25, 71.2833, 7.925, 53.1, 8.05, 8.4583, 51.8625, 21.075, 11.1333, 30.0708, 16.7, 26.55, 8.05, 31.275, 7.8542, 16.0, 29.125, 13.0, 18.0, 7.225, 26.0, 13.0, 8.0292, 35.5, 21.075, 31.3875, 7.225, 263.0, 7.8792, 7.8958, 27.7208, 146.5208, 7.75, 10.5, 82.1708, 52.0, 7.2292, 8.05, 18.0, 11.2417, 9.475, 21.0, 7.8958, 41.5792, 7.8792, 8.05, 15.5, 7.75, 21.6792, 17.8, 39.6875, 7.8, 76.7292, 26.0, 61.9792, 35.5, 10.5, 7.2292, 27.75, 46.9, 7.2292, 80.0, 83.475, 27.9, 27.7208, 15.2458, 10.5, 8.1583, 7.925, 8.6625, 10.5, 46.9, 73.5, 14.4542, 56.4958, 7.65, 7.8958, 8.05, 29.0, 12.475, 9.0, 9.5, 7.7875, 47.1, 10.5, 15.85, 34.375, 8.05, 263.0, 8.05, 8.05, 7.8542, 61.175, 20.575, 7.25, 8.05, 34.6542, 63.3583, 23.0, 26.0, ... ]

### 컬럼 여러개 선택하기

In [57]:
train_data['Fare', 'Embarked']

Columns:
	Fare	float
	Embarked	str

Rows: 891

Data:
+---------+----------+
|   Fare  | Embarked |
+---------+----------+
|   7.25  |    S     |
| 71.2833 |    C     |
|  7.925  |    S     |
|   53.1  |    S     |
|   8.05  |    S     |
|  8.4583 |    Q     |
| 51.8625 |    S     |
|  21.075 |    S     |
| 11.1333 |    S     |
| 30.0708 |    C     |
+---------+----------+
[891 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

# 비어 있는 값 찾기

- NA (None) 값이 있는 경우
- string type 데이터에서 값이 ''(공백)으로 되어 있는 경우

### None으로 되어 있는 경우

In [77]:
test_data[test_data['Fare'] == None]

Columns:
	PassengerId	int
	Pclass	int
	Name	str
	Sex	str
	Age	float
	SibSp	int
	Parch	int
	Ticket	str
	Fare	float
	Cabin	str
	Embarked	str

Rows: Unknown

Data:
+-------------+--------+--------------------+------+------+-------+-------+--------+
| PassengerId | Pclass |        Name        | Sex  | Age  | SibSp | Parch | Ticket |
+-------------+--------+--------------------+------+------+-------+-------+--------+
|     1044    |   3    | Storey, Mr. Thomas | male | 60.5 |   0   |   0   |  3701  |
+-------------+--------+--------------------+------+------+-------+-------+--------+
+------+-------+----------+
| Fare | Cabin | Embarked |
+------+-------+----------+
| None |       |    S     |
+------+-------+----------+
[? rows x 11 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [34]:
fare_for_none = test_data['Fare'].mean()

fillna 메소드를 이용해서 None으로 비어 있는 값을 지정할 수 있습니다.

    SFrame.fillna(column=컬럼이름, value=지정하고자하는값)
    
**주의: 반드시 다시 SFrame에 저장해야 합니다.** 여기서는 **new_test_data**라는 새로운 변수명을 사용합시다.

In [35]:
new_test_data = test_data.fillna(column='Fare', value=fare_for_none)

In [36]:
new_test_data[new_test_data['PassengerId'] == 1044]

Columns:
	PassengerId	int
	Pclass	int
	Name	str
	Sex	str
	Age	float
	SibSp	int
	Parch	int
	Ticket	str
	Fare	float
	Cabin	str
	Embarked	str

Rows: Unknown

Data:
+-------------+--------+--------------------+------+------+-------+-------+--------+
| PassengerId | Pclass |        Name        | Sex  | Age  | SibSp | Parch | Ticket |
+-------------+--------+--------------------+------+------+-------+-------+--------+
|     1044    |   3    | Storey, Mr. Thomas | male | 60.5 |   0   |   0   |  3701  |
+-------------+--------+--------------------+------+------+-------+-------+--------+
+---------------+-------+----------+
|      Fare     | Cabin | Embarked |
+---------------+-------+----------+
| 35.6271884892 |       |    S     |
+---------------+-------+----------+
[? rows x 11 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

### 공백으로 되어 있는 경우

In [49]:
train_data[train_data['Embarked'] == '']

Columns:
	PassengerId	int
	Survived	int
	Pclass	int
	Name	str
	Sex	str
	Age	float
	SibSp	int
	Parch	int
	Ticket	str
	Fare	float
	Cabin	str
	Embarked	str

Rows: Unknown

Data:
+-------------+----------+--------+-------------------------------+--------+
| PassengerId | Survived | Pclass |              Name             |  Sex   |
+-------------+----------+--------+-------------------------------+--------+
|      62     |    1     |   1    |      Icard, Miss. Amelie      | female |
|     830     |    1     |   1    | Stone, Mrs. George Nelson ... | female |
+-------------+----------+--------+-------------------------------+--------+
+------+-------+-------+--------+------+-------+----------+
| Age  | SibSp | Parch | Ticket | Fare | Cabin | Embarked |
+------+-------+-------+--------+------+-------+----------+
| 38.0 |   0   |   0   | 113572 | 80.0 |  B28  |          |
| 62.0 |   0   |   0   | 113572 | 80.0 |  B28  |          |
+------+-------+-------+--------+------+-------+----------+
[? rows x 12 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

이 값은 다음 섹션에서 함수를 만들어서 채우고자 합니다.

### numpy는 isnull을 사용합니다.

pandas와 numpy를 사용하는 경우 다음 메소드를 이용하면 각 셀에 들어있는 데이터에 null 값이 있는지를 확인해 줍니다. 

    DataFrame.isnulll()
    Series.isnull()

In [94]:
import numpy as np
import pandas as pd

train_df = pd.read_csv('../data/kaggle_titanic/train.csv')
test_df = pd.read_csv('../data/kaggle_titanic/test.csv')

In [83]:
test_df.isnull()

PassengerId  Pclass   Name    Sex    Age  SibSp  Parch  Ticket   Fare  \
0          False   False  False  False  False  False  False   False  False   
1          False   False  False  False  False  False  False   False  False   
2          False   False  False  False  False  False  False   False  False   
3          False   False  False  False  False  False  False   False  False   
4          False   False  False  False  False  False  False   False  False   
5          False   False  False  False  False  False  False   False  False   
6          False   False  False  False  False  False  False   False  False   
7          False   False  False  False  False  False  False   False  False   
8          False   False  False  False  False  False  False   False  False   
9          False   False  False  False  False  False  False   False  False   
10         False   False  False  False   True  False  False   False  False   
11         False   False  False  False  False  False  False   False  False   
12         False   False  False  False  False  False  False   False  False   
13         False   False  False  False  False  False  False   False  False   
14         False   False  False  False  False  False  False   False  False   
15         False   False  False  False  False  False  False   False  False   
16         False   False  False  False  False  False  False   False  False   
17         False   False  False  False  False  False  False   False  False   
18         False   False  False  False  False  False  False   False  False   
19         False   False  False  False  False  False  False   False  False   
20         False   False  False  False  False  False  False   False  False   
21         False   False  False  False  False  False  False   False  False   
22         False   False  False  False   True  False  False   False  False   
23         False   False  False  False  False  False  False   False  False   
24         False   False  False  False  False  False  False   False  False   
25         False   False  False  False  False  False  False   False  False   
26         False   False  False  False  False  False  False   False  False   
27         False   False  False  False  False  False  False   False  False   
28         False   False  False  False  False  False  False   False  False   
29         False   False  False  False   True  False  False   False  False   
..           ...     ...    ...    ...    ...    ...    ...     ...    ...   
388        False   False  False  False  False  False  False   False  False   
389        False   False  False  False  False  False  False   False  False   
390        False   False  False  False  False  False  False   False  False   
391        False   False  False  False  False  False  False   False  False   
392        False   False  False  False  False  False  False   False  False   
393        False   False  False  False  False  False  False   False  False   
394        False   False  False  False  False  False  False   False  False   
395        False   False  False  False  False  False  False   False  False   
396        False   False  False  False  False  False  False   False  False   
397        False   False  False  False  False  False  False   False  False   
398        False   False  False  False  False  False  False   False  False   
399        False   False  False  False  False  False  False   False  False   
400        False   False  False  False  False  False  False   False  False   
401        False   False  False  False  False  False  False   False  False   
402        False   False  False  False  False  False  False   False  False   
403        False   False  False  False  False  False  False   False  False   
404        False   False  False  False  False  False  False   False  False   
405        False   False  False  False  False  False  False   False  False   
406        False   False  False  False  False  False  False   False  False   
407        False   False  F

In [84]:
test_df['Fare'].isnull()

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19     False
20     False
21     False
22     False
23     False
24     False
25     False
26     False
27     False
28     False
29     False
       ...  
388    False
389    False
390    False
391    False
392    False
393    False
394    False
395    False
396    False
397    False
398    False
399    False
400    False
401    False
402    False
403    False
404    False
405    False
406    False
407    False
408    False
409    False
410    False
411    False
412    False
413    False
414    False
415    False
416    False
417    False
Name: Fare, Length: 418, dtype: bool

### isnull은 numeric data type인 경우 NaN을 찾아 줍니다.

In [92]:
test_df[test_df['Fare'].isnull()]

PassengerId  Pclass                Name   Sex   Age  SibSp  Parch Ticket  \
152         1044       3  Storey, Mr. Thomas  male  60.5      0      0   3701   

     Fare Cabin Embarked  
152   NaN   NaN        S

### isnull은 string data type인 경우 공백을 찾아 줍니다.

In [110]:
train_df[train_df['Embarked'].isnull()]

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

# 함수 만들기

    def 함수명(받고자하는인자1, 받고자하는인자2, ...):
        
- 인자는 없을 수도 있다.
- 반드시 뒤에 :을 사용하자
- 값을 반환하려면 return을 사용하자

In [40]:
def fillSforEmbarked(value):
    if value == '':
        return 'S'
    else:
        return value

In [50]:
train_data['Embarked_filled'] = train_data['Embarked'].apply(fillSforEmbarked)

In [51]:
train_data[(train_data['PassengerId'] == 62) | (train_data['PassengerId'] == 830)]

Columns:
	PassengerId	int
	Survived	int
	Pclass	int
	Name	str
	Sex	str
	Age	float
	SibSp	int
	Parch	int
	Ticket	str
	Fare	float
	Cabin	str
	Embarked	str
	Embarked_filled	str

Rows: Unknown

Data:
+-------------+----------+--------+-------------------------------+--------+
| PassengerId | Survived | Pclass |              Name             |  Sex   |
+-------------+----------+--------+-------------------------------+--------+
|      62     |    1     |   1    |      Icard, Miss. Amelie      | female |
|     830     |    1     |   1    | Stone, Mrs. George Nelson ... | female |
+-------------+----------+--------+-------------------------------+--------+
+------+-------+-------+--------+------+-------+----------+-----------------+
| Age  | SibSp | Parch | Ticket | Fare | Cabin | Embarked | Embarked_filled |
+------+-------+-------+--------+------+-------+----------+-----------------+
| 38.0 |   0   |   0   | 113572 | 80.0 |  B28  |          |        S        |
| 62.0 |   0   |   0   | 113572 | 80.0 |  B28  |          |        S        |
+------+-------+-------+--------+------+-------+----------+-----------------+
[? rows x 13 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

# 시간 측정하기

- 시간을 측정해서 더 나은 방법을 찾는다.
- 먼저 작은 데이터에 대해서 테스트를 진행하고 큰 데이터를 적용해 보자.

In [1]:
import time

start_time = time.time()
count = 0
for i in range(100000):
    count += 1
print count

end_time = time.time()

print "duration: ", end_time - start_time

100000
duration:  0.041435956955


In [159]:
products = gl.SFrame('../data/amazon_baby.gl/')
products['word_count'] = gl.text_analytics.count_words(products['review'])
filtered_products = products[products['rating'] != 3]

selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

def selected_word_count(dict, w):
    if w in dict:
        return dict[w]
    else:
        return 0

### loop로 직접 세기 BOMSHIK's solution

In [21]:
start_time = time.time()

temp_wc = filtered_products['word_count']
temp_len = len (temp_wc)

for w in selected_words:
    i = 0 
    cnt = 0
    while i < temp_len :
        cnt = cnt + selected_word_count(temp_wc[i], w)
        i = i+1
    print w + " : " + str(cnt)
    
end_time = time.time()
print "duration: ", end_time - start_time

awesome : 2002
great : 42420
fantastic : 873
amazing : 1305
love : 40277
horrible : 659
bad : 3197
terrible : 673
awful : 345
wow : 131
hate : 1057
duration:  369.424745083


이 솔루션은 369.4초가 걸렸습니다.

### lambda와 apply 사용하기

In [20]:
start_time = time.time()

for w in selected_words:
    filtered_products[w] = filtered_products['word_count'].apply(lambda x:selected_word_count(x, w))
    print w, filtered_products[w].sum()
    
end_time = time.time()
print "duration: ", end_time - start_time

awesome 2002
great 42420
fantastic 873
amazing 1305
love 40277
horrible 659
bad 3197
terrible 673
awful 345
wow 131
hate 1057
duration:  55.6740069389


이 솔루션은 55.6초가 걸렸습니다.

### 결론
 
 - length를 계산하는 게 불필요하다.
 - apply의 강력함을 알 수 있다.
 - 카운트를 개별적으로 세는 것보다 SArray의 sum으로 계산하는 게 훨씬 빠르다.
 
**데이터가 많을수록 효과는 극적으로 벌어진다. 따라서 apply와 sum을 잘 활용하자.**

# pandas의 get_dummies

카테고리 값이 2개 이상인 feature들을 각각의 개별 컬럼으로 만들어 줍시다. 해당 결과를 **X_dummies**에 저장합시다.

    pd.get_dummies(데이터셋, 변환하고자하는컬럼이름들리스트, drop_first=True)
    

In [120]:
full_df = pd.concat([train_df.drop('Survived', axis=1), test_df], axis=0)

In [121]:
full_df.head()

PassengerId  Pclass                                               Name  \
0            1       3                            Braund, Mr. Owen Harris   
1            2       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2            3       3                             Heikkinen, Miss. Laina   
3            4       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4            5       3                           Allen, Mr. William Henry   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S  
1  female  38.0      1      0          PC 17599  71.2833   C85        C  
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3  female  35.0      1      0            113803  53.1000  C123        S  
4    male  35.0      0      0            373450   8.0500   NaN        S

In [122]:
converting_columns = ['Pclass']
drop_features = ['PassengerId', 'Name', 'Ticket', 'Cabin', 'Embarked', 'Sex', 'Age']
features = list(set(full_df.columns.values.tolist()) - set(drop_features))

In [123]:
X_dummies = pd.get_dummies(full_df[features], 
                           columns = converting_columns,
                           drop_first=True)

In [124]:
X_dummies.head()

Fare  SibSp  Parch  Pclass_2  Pclass_3
0   7.2500      1      0         0         1
1  71.2833      1      0         0         0
2   7.9250      0      0         0         1
3  53.1000      1      0         0         0
4   8.0500      0      0         0         1

### 결론

drop_first는 해당값이 0인 컬럼을 별도로 생성하지 않는다.

# Image data & Normalize

이미지는 가로 x 세로 x 채널수 만큼의 수 - 픽셀의 수 - 로 나타낼 수 있다. 채널수는 단색은 1, RGB는 3이다.

한 픽셀은 색상을 몇가지로 표현할지에 따라서 다양한 값을 가질 수 있다. 256bit인 경우 0 ~ 255의 값을 가진다.


### normarlization
특정 구간의 값을 0 ~ 1 사이로 만든다. 0 ~ 255의 구간에서 변동하는 값이라면 255로 나누면 된다.

In [126]:
image_train = gl.SFrame('../data/image_train_data/')

In [132]:
image_train.head()

Columns:
	id	int
	image	Image
	label	str
	deep_features	array
	image_array	array

Rows: 10

Data:
+-----+----------------------+------------+-------------------------------+
|  id |        image         |   label    |         deep_features         |
+-----+----------------------+------------+-------------------------------+
|  24 | Height: 32 Width: 32 |    bird    | [0.242871761322, 1.0954537... |
|  33 | Height: 32 Width: 32 |    cat     | [0.525087952614, 0.0, 0.0,... |
|  36 | Height: 32 Width: 32 |    cat     | [0.566015958786, 0.0, 0.0,... |
|  70 | Height: 32 Width: 32 |    dog     | [1.12979578972, 0.0, 0.0, ... |
|  90 | Height: 32 Width: 32 |    bird    | [1.71786928177, 0.0, 0.0, ... |
|  97 | Height: 32 Width: 32 | automobile | [1.57818555832, 0.0, 0.0, ... |
| 107 | Height: 32 Width: 32 |    dog     | [0.0, 0.0, 0.220677852631,... |
| 121 | Height: 32 Width: 32 |    bird    | [0.0, 0.23753464222, 0.0, ... |
| 136 | Height: 32 Width: 32 | automobile | [0.0, 0.0, 0.0, 0.0, 0.0, ... |
| 138 | Height: 32 Width: 32 |    bird    | [0.658935725689, 0.0, 0.0,... |
+-----+----------------------+------------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [73.0, 77.0, 58.0, 71.0, 6... |
| [7.0, 5.0, 8.0, 7.0, 5.0, ... |
| [169.0, 122.0, 65.0, 131.0... |
| [154.0, 179.0, 152.0, 159.... |
| [216.0, 195.0, 180.0, 201.... |
| [33.0, 44.0, 27.0, 29.0, 4... |
| [97.0, 51.0, 31.0, 104.0, ... |
| [93.0, 96.0, 88.0, 102.0, ... |
| [35.0, 59.0, 53.0, 36.0, 5... |
| [205.0, 193.0, 195.0, 200.... |
+-------------------------------+
[10 rows x 5 columns]

In [135]:
type(image_train['image_array'])

graphlab.data_structures.sarray.SArray

In [136]:
image_train['image_array'][0]

array('d', [73.0, 77.0, 58.0, 71.0, 68.0, 50.0, 77.0, 69.0, 44.0, 120.0, 116.0, 83.0, 125.0, 120.0, 90.0, 155.0, 148.0, 117.0, 147.0, 152.0, 106.0, 133.0, 136.0, 92.0, 112.0, 101.0, 75.0, 100.0, 84.0, 68.0, 100.0, 78.0, 74.0, 69.0, 49.0, 51.0, 22.0, 16.0, 13.0, 35.0, 27.0, 24.0, 52.0, 40.0, 40.0, 70.0, 55.0, 54.0, 117.0, 97.0, 89.0, 122.0, 101.0, 99.0, 103.0, 85.0, 88.0, 88.0, 74.0, 73.0, 68.0, 59.0, 52.0, 72.0, 65.0, 52.0, 82.0, 73.0, 60.0, 84.0, 72.0, 68.0, 84.0, 67.0, 73.0, 77.0, 56.0, 65.0, 93.0, 75.0, 81.0, 117.0, 102.0, 98.0, 174.0, 171.0, 146.0, 183.0, 210.0, 149.0, 180.0, 214.0, 143.0, 185.0, 225.0, 144.0, 73.0, 74.0, 52.0, 75.0, 68.0, 55.0, 78.0, 64.0, 50.0, 103.0, 90.0, 64.0, 80.0, 65.0, 35.0, 127.0, 110.0, 80.0, 136.0, 127.0, 98.0, 141.0, 134.0, 109.0, 129.0, 113.0, 92.0, 115.0, 97.0, 72.0, 111.0, 91.0, 61.0, 115.0, 96.0, 67.0, 94.0, 83.0, 65.0, 92.0, 78.0, 65.0, 78.0, 60.0, 53.0, 89.0, 68.0, 64.0, 123.0, 99.0, 92.0, 117.0, 93.0, 92.0, 118.0, 97.0, 102.0, 121.0, 103.0, 103.0

In [138]:
len(image_train['image_array'][0])

3072

In [139]:
type(image_train['image_array'][0])

array.array

In [140]:
32*32*3

3072

32x32 픽셀 크기의 RGB 컬러 이미지이기 때문에 32x32x3 = 3072만큼의 데이터가 필요합니다. 각 픽셀은 256bit 그라디에이션 이미지이기 때문에 0~255의 값을 가집니다.

In [143]:
image_train['image_array'][0][0:32]

array('d', [73.0, 77.0, 58.0, 71.0, 68.0, 50.0, 77.0, 69.0, 44.0, 120.0, 116.0, 83.0, 125.0, 120.0, 90.0, 155.0, 148.0, 117.0, 147.0, 152.0, 106.0, 133.0, 136.0, 92.0, 112.0, 101.0, 75.0, 100.0, 84.0, 68.0, 100.0, 78.0])

첫번째 이미지의 R 색상의 맨 윗줄의 데이터가 출력됩니다.

In [153]:
def normalize_pixel(array):
    new_list = []
    for v in array:
        new_list.append(v/255)
    return new_list

In [154]:
image_train_norm = image_train['image_array'][0:1].apply(normalize_pixel)

In [155]:
image_train_norm[0][0:32]

array('d', [0.28627450980392155, 0.30196078431372547, 0.22745098039215686, 0.2784313725490196, 0.26666666666666666, 0.19607843137254902, 0.30196078431372547, 0.27058823529411763, 0.17254901960784313, 0.47058823529411764, 0.4549019607843137, 0.3254901960784314, 0.49019607843137253, 0.47058823529411764, 0.35294117647058826, 0.6078431372549019, 0.5803921568627451, 0.4588235294117647, 0.5764705882352941, 0.596078431372549, 0.41568627450980394, 0.5215686274509804, 0.5333333333333333, 0.3607843137254902, 0.4392156862745098, 0.396078431372549, 0.29411764705882354, 0.39215686274509803, 0.32941176470588235, 0.26666666666666666, 0.39215686274509803, 0.3058823529411765])

In [158]:
image_train['image_array'][0][0] / 255

0.28627450980392155